### PR TITLE
iOS Universal Links and Android App Links

### PR DESCRIPTION
**Pull request by cdanwards**
### What does this PR do, or why is it needed?

This PR adds the native code, id parser and linking handling for ios universal links and android app links.

This allows phones that have the app downloaded to point a `newspring.cc` link to that piece of content in the app.

The corresponding association files for the website can be found at:
```
newspring.cc/.well-known/assetlink.json
newspring.cc/apple-app-site-association
```

### How do I test this PR?
Load an android sim and run:
`adb shell am start -W -a android.intent.action.VIEW -d "https://newspring.cc/articles/how-to-help-your-young-adult-manage-money-wisely`
Load an apple sim and run:
`xcrun simctl openurl booted https://newspring.cc/articles/how-to-help-your-young-adult-manage-money-wisely`

## TODO

- [ ] I am affirming this is my _best_ work ([Ecclesiastes 9:10](https://www.bible.com/bible/97/ECC.9.10.MSG))
- [ ] PR has a relevant title that will be understandable in a public changelog (ie...non developers)
- [ ] No new warnings in tests, in storybook, and in-app
- [ ] Upload GIF(s) of iOS and Android if applicable
- [ ] Set a relevant reviewer

## REVIEW

- [ ] Review updates to test coverage and snapshots
- [ ] Review code through the lens of being concise, simple, and well-documented

**Manual QA**

- [ ] Manual QA on iOS and ensure it looks/behaves as expected
- [ ] Manual QA on Android and ensure it looks/behaves as expected

---

> The purpose of PR Review is to _improve the quality of the software._

### CODE REVIEW

**Comment by cdanwards** [comment]
@redreceipt @richarddubay One of yall should make sure to run tests. My local snaps are somehow messed up.

File: packages/apollos-church-api/src/data/content-items/data-source.js [hunk 1]
```diff
@@ -203,4 +203,16 @@ export default class ContentItem extends oldContentItem.dataSource {
     });
     return features;
   }
+
+  getBySlug = async (slug) => {
+    const contentItemFromSlug = await this.request('ContentChannelItemSlugs')
+      .filter(`Slug eq '${slug}'`)
+      .get();
```
Comment (redreceipt) [inline packages/apollos-church-api/src/data/content-items/data-source.js]:
```suggestion
      .first();
```

File: packages/apollos-church-api/src/data/content-items/data-source.js [hunk 2]
```diff
@@ -203,4 +203,16 @@ export default class ContentItem extends oldContentItem.dataSource {
...
+    console.log(slug, contentItemFromSlug);
```
Comment (redreceipt) [inline packages/apollos-church-api/src/data/content-items/data-source.js]:
```suggestion
```

File: packages/apollos-church-api/src/data/content-items/data-source.js [hunk 3]
```diff
@@ -203,4 +203,16 @@ export default class ContentItem extends oldContentItem.dataSource {
...
+    if (contentItemFromSlug === [])
```
Comment (redreceipt) [inline packages/apollos-church-api/src/data/content-items/data-source.js]:
```suggestion
    if (slug)
```

File: packages/apollos-church-api/src/data/content-items/data-source.js [hunk 4]
```diff
@@ -203,4 +203,16 @@ export default class ContentItem extends oldContentItem.dataSource {
     });
     return features;
   }
+
+  getBySlug = async (slug) => {
+    const contentItemFromSlug = await this.request('ContentChannelItemSlugs')
```
Comment (redreceipt) [inline packages/apollos-church-api/src/data/content-items/data-source.js]:
```suggestion
    const slug = await this.request('ContentChannelItemSlugs')
```

File: packages/apollos-church-api/src/data/content-items/data-source.js [hunk 5]
```diff
@@ -203,4 +203,16 @@ export default class ContentItem extends oldContentItem.dataSource {
...
+      .filter(`Slug eq '${slug}'`)
+      .get();
+    console.log(slug, contentItemFromSlug);
+    if (contentItemFromSlug === [])
+      throw new Error('There is no Content Item with the slug provided.');
```
Comment (redreceipt) [inline packages/apollos-church-api/src/data/content-items/data-source.js]:
```suggestion
      throw new Error('Slug does not exist.');
```

File: packages/apollos-church-api/src/data/content-items/data-source.js [hunk 6]
```diff
@@ -203,4 +203,16 @@ export default class ContentItem extends oldContentItem.dataSource {
...
+    console.log(slug, contentItemFromSlug);
```
Comment (redreceipt) [inline packages/apollos-church-api/src/data/content-items/data-source.js]:
```suggestion
```

File: packages/apollos-church-api/src/data/content-items/data-source.js [hunk 7]
```diff
@@ -203,4 +203,16 @@ export default class ContentItem extends oldContentItem.dataSource {
...
+
+    return this.getFromId(`${contentItemFromSlug[0].contentChannelItemId}`);
```
Comment (redreceipt) [inline packages/apollos-church-api/src/data/content-items/data-source.js]:
```suggestion
    return this.getFromId(`${slug.contentChannelItemId}`);
```

File: packages/newspringchurchapp/src/linking/Provider.js [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+          return queryResult;
+        });
+      const { data, loading } = res;
+
+      if (loading) return null;
+
+      return data;
+    };
+    const queryResult = await slugQuery();
+    console.log(queryResult.contentItemFromSlug.id);
```
Comment (redreceipt) [inline packages/newspringchurchapp/src/linking/Provider.js]:
```suggestion
```

File: packages/newspringchurchapp/src/linking/Provider.js [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    const newUrl = `newspringchurchapp://AppStackNavigator/ContentSingle?itemId=${
+      queryResult.contentItemFromSlug.id
+    }`;
+    console.log(newUrl);
```
Comment (redreceipt) [inline packages/newspringchurchapp/src/linking/Provider.js]:
```suggestion
```

**Review by redreceipt — CHANGES_REQUESTED** [review]

**Comment by redreceipt** [comment]
Nice work! Want to add a test for the new query and new data source function?

File: packages/apollos-church-api/src/data/content-items/data-source.js (hunk 1 above)
Comment (cdanwards) [reply packages/apollos-church-api/src/data/content-items/data-source.js]: This is nifty. I forgot that this method existed.

File: packages/newspringchurchapp/src/linking/Provider.js [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+      if (loading) return null;
+
+      return data;
+    };
+    const queryResult = await slugQuery();
+    const newUrl = `newspringchurchapp://AppStackNavigator/ContentSingle?itemId=${
+      queryResult.contentItemFromSlug.id
+    }`;
+    if (newUrl) {
```
Comment (redreceipt) [inline packages/newspringchurchapp/src/linking/Provider.js]: I don't think this `if` statement is doing anything. Creates an unnecessary test coverage gap.

File: packages/apollos-church-api/src/data/content-items/data-source.js [hunk 11]
```diff
@@ -203,4 +203,13 @@ export default class ContentItem extends oldContentItem.dataSource {
     });
     return features;
   }
+
+  getBySlug = async (slug) => {
+    const contentItem = await this.request('ContentChannelItemSlugs')
```
Comment (redreceipt) [inline packages/apollos-church-api/src/data/content-items/data-source.js]:
```suggestion
    const contentItemSlug = await this.request('ContentChannelItemSlugs')
```

**Review by redreceipt — CHANGES_REQUESTED** [review]

File: packages/apollos-church-api/src/data/content-items/data-source.js (hunk 11 above)
Comment (cdanwards) [reply packages/apollos-church-api/src/data/content-items/data-source.js]: Should this just be the original contentItemFromSlug? haha

Comment (redreceipt) [reply packages/apollos-church-api/src/data/content-items/data-source.js]: it's not a content item table, it's a slug table

**Review by redreceipt — APPROVED** [review]

**Review by richarddubay — APPROVED** [review]
👍

**Pull request merged**